### PR TITLE
feat!: added `PolylinePattern` with support for `solid`, `dotted`, `dashed` styles

### PR DESCRIPTION
**Pull request by monsieurtanuki**
Impacted files:
* `pages/polyline.dart`: added a dashed polyline as example
* `polyline_layer/polyline.dart`: added the `dashValues` field for `Polyline`
* `painter.dart`: now handling the dashed polylines

Screenshot (new dashed polyline between Paris and Nice):
Screenshot_1710696376

What now?
* the use of `isDotted` / `segmentSpacingFactor` / `dashValues` fields is a bit clumsy
=> we would be better off with a unique field
=> new class `PolylineStyle` with "plain", "dotted(space)", "dashed(dashvalues)"?
* we could reuse for the dotted style the algorithm implemented in the new method `getDistanceOffset` used for "dashed style" (something like: give me the offset of the point on this [A,B] segment distant from A by n pixels)
* for the moment the "final point situation" is not dealt with: what if the latest point of the dashed line is invisible? We should explicitly put a dot here, shouldn't we?

### CODE REVIEW

**Comment by JaffaKetchup** [comment]
Hey, this looks great! Huge thanks for ticking this off the list.

> the use of isDotted / segmentSpacingFactor / dashValues fields is a bit clumsy
=> we would be better off with a unique field
=> new class PolylineStyle with "plain", "dotted(space)", "dashed(dashvalues)"?

Yes, I agree with this. I've added a `PolylinePattern` object which does this. It's a little bit of complex logic internally, but it keeps the externals small.

> we could reuse for the dotted style the algorithm implemented in the new method getDistanceOffset used for "dashed style" (something like: give me the offset of the point on this [A,B] segment distant from A by n pixels)

We generally try to avoid including these generic tools publicly, the major exception being the simplification tools. For now, I'm happy to keep this internal, and we can always make it public in future.

> for the moment the "final point situation" is not dealt with: what if the latest point of the dashed line is invisible? We should explicitly put a dot here, shouldn't we?

Yep, we should deal with this. I'm not quite sure the best way to do this, so if you have something in mind, please go ahead!

File: lib/src/layer/polyline_layer/pattern.dart
Comment (JaffaKetchup) [reply lib/src/layer/polyline_layer/pattern.dart:33]: I think all negative numbers are less than 2 ;)

Comment (JaffaKetchup) [reply lib/src/layer/polyline_layer/pattern.dart:33]: Ah, my apologies, misunderstood. Checking values is a little bit more expensive than the length though. It should be ok, I've included it in the docstring.

File: lib/src/layer/polyline_layer/painter.dart
Comment (JaffaKetchup) [reply lib/src/layer/polyline_layer/painter.dart:165]: > There is no "dotted" without spacingFactor, as it defaults to 1.5 in PolylinePattern.

If I understand you correctly, I think you've made an incorrect assumption. `spacingFactor` only defaults to 1.5 when using `PolylinePattern.dotted`, otherwise it defaults to `null`.

This deprecations allows `PolylinePattern` to be `solid` (as is now default), but dotting to be enabled with a spacing factor of 1.5 when `isDotted` is set `true`.

File: lib/src/layer/polyline_layer/polyline.dart
Comment (JaffaKetchup) [reply lib/src/layer/polyline_layer/polyline.dart:75]: `segmentSpacingFactor` was not deprecated, only `isDotted` was.

File: lib/src/layer/polyline_layer/painter.dart
Comment (monsieurtanuki) [reply lib/src/layer/polyline_layer/painter.dart:165]: What I mean is that `spacingFactor` is not null if and only if we used the `dotted` constructor.
Therefore it's like
```dart
bool get isDotted => spacingFactor != null;
```

File: lib/src/layer/polyline_layer/polyline.dart
Comment (JaffaKetchup) [reply lib/src/layer/polyline_layer/polyline.dart:64]: Yeah, I think that method of deprecation is better, seeing as `Polyline` is non-`const`. I've made that change.

File: lib/src/layer/polyline_layer/painter.dart
Comment (JaffaKetchup) [reply lib/src/layer/polyline_layer/painter.dart]: Changed the deprecation method as you suggested, makes more sense now.

File: lib/src/layer/polyline_layer/painter.dart
Comment (JaffaKetchup) [reply lib/src/layer/polyline_layer/painter.dart]: Changed the deprecation method as you suggested, makes more sense now.

File: lib/src/layer/polyline_layer/painter.dart
Comment (monsieurtanuki) [inline lib/src/layer/polyline_layer/painter.dart:266]:
```suggestion
    if (dashValues.length < 2) {
      return null;
    }
```

**Review by monsieurtanuki — COMMENTED** [review]
@JaffaKetchup Looks good to me!

Still pending: what to do with the last point of a dashed line? I suggest a dot - would you like me to work on it? (I could start only this week-end).

And the "if solid / if dashed / if dotted" sequence is a bit clumsy but it's not very important - except that we have similar algorithms coded twice for dashed and dotted.

Not important suggestion: we could make a dot part of a dashed line, like Morse code (dash-space-dot-space-dash...)

**Comment by monsieurtanuki** [comment]
@JaffaKetchup I've just fixed the side effect problem:
```dart
  /// The idea is that we need to be able to display the dash pattern completely
  /// n times (at least once), plus once the initial dash segment. That's the
  /// way we deal with the "ending" side-effect.
```

| [500, 500, 300, 1000] | [50, 50, 30, 100] | [50, 20, 25, 20] |
| -- | -- | -- |
| Screenshot_1711179486 | Screenshot_1711179531 | Screenshot_1711179596 |

**Review by JaffaKetchup — DISMISSED** [review]
LGTM, thanks! Not entirely sure what you did in that last bit, but it seems to have worked and looks nice.

Any other maintainers want to have a review?

**Comment by monsieurtanuki** [comment]
> LGTM, thanks! Not entirely sure what you did in that last bit, but it seems to have worked and looks nice.

@JaffaKetchup As explained, there's a side-effect with dash pattern - as there is obviously with dotted pattern: if the dash pattern is 50+20 pixels, what do we do if the polyline length is 60 pixels? Just display the first 50 pixels and nothing else, so that the last point (presumably very important) is not displayed at all?
The suggestion implemented here is to always display "(dash pattern) * n + first dash segment" where n > 0. So we always display the first point and the last point.

**Comment by monsieurtanuki** [comment]
Still playing with side effects, with dash segments `[50, 20, 100, 100]`:

| don't see the final point (Nice) | see Nice as a mere dot | used a factor on segment values |
| -- | -- | -- |
| Screenshot_1711356063 | Screenshot_1711355991 | Screenshot_1711356103 |

1. if we don't handle side-effects, we may not see the final point
2. handling the side-effect - display the last point as a dot if needed
3. handling the side-effect - compute and use a factor so that we display an integer number of times the segments, plus once the first segment, so that the last point is always displayed

Actually I prefer solution 2, because solution 1 is problematic (no final point) and solution 3 is hard to explain anyway.

@JaffaKetchup
* Would you like me to implement solution 2?
* Would you like me to implement something similar for dotted style?
* I think the current PR has been re-labeled as "breaking change" (`feat!`), which is not really, is it?

**Comment by JaffaKetchup** [comment]
Hey, thanks for continuing to work on this.

Option one is unnacceptable, but I feel like option 2 and 3 are both valid. Additionally, rather than option 2, if there are enough dashes, we could do it so that it just dashes as normal, then truncates short the last dash (if it isn't a gap). Not sure how difficult this could be to implement. If you've got time, and it's not too complex, I guess allowing the user to pick between them all with some sort of `PatternFit` option would be good. But I don't want to exhaust you.

For dotted lines, I guess only option 3 makes sense (scaling the factor), because otherwise it will look a bit strange? Again, `PatternFit` could work.

The PR is breaking because it removes the `isDotted` option.

**Comment by monsieurtanuki** [comment]
@JaffaKetchup I've just introduced the new `PatternFit` field, for both dotted and dashed styles.

| x | dashed [100, 40] | dotted (10) |
| -- | -- | -- |
| resize | Screenshot_1711463836 | Screenshot_1711463724 |
| lastPoint | Screenshot_1711463786 | Screenshot_1711463714 |

**Comment by JaffaKetchup** [comment]
@monsieurtanuki I've run with those changes (thanks for those :)) with a couple more options, just so we can give users every option. It doesn't add too much complexity. Let me know what you think!

File: lib/src/layer/polyline_layer/painter.dart
Comment (monsieurtanuki) [reply lib/src/layer/polyline_layer/painter.dart:413]: My latest suggestion is `_PixelHiker` and `_doneWithCurrentOffset`, but I'm open to other names.

**Comment by monsieurtanuki** [comment]
> @monsieurtanuki I've run with those changes (thanks for those :)) with a couple more options, just so we can give users every option. It doesn't add too much complexity. Let me know what you think!

@JaffaKetchup Thank you for your rewritings! My comments...

I'm not convinced with `PatternFit` having "forbidden" values, like `extendFinalDash` not available for `dotted` polylines. I would have created separate `DottedPatternFit` and `DashedPatternFit` enums instead. But that's not very important.

I don't believe your `extendFinalDash` works - or I haven't understood the goal.
My understanding was that instead of displaying spaces for the very last invisible parts, you were displaying lines. But the way you coded it you display only the very last _offset_ segment: from the latest `Offset`, but not following the route from the latest `lineTo`.

* (`---  ---  ---  `) as `none`
* (`---  ---  -----`) as `extendFinalDash` as I've understood it
* (`---  ---  --- -`) as `extendFinalDash` as it's currently coded, potentially

For `extendFinalDash`, I would expect a final zigzag, not a straight line.
| `none` | `scale` | `extendFinalDash` |
| -- | -- | -- |
| Screenshot_1711528964 | Screenshot_1711528982 | Screenshot_1711528996 |

Working on it...

**Comment by monsieurtanuki** [comment]
@JaffaKetchup Latest push:
* `PatternFit.none`
* `PatternFit.appendDot`
* `PatternFit.extendFinalDash`
* `PatternFit.scaleUp`
* `PatternFit.scaleDown`

| dotted | dashed |
| -- | -- |
| Screenshot_1711535174 | Screenshot_1711534256 |

**Review by JaffaKetchup — APPROVED** [review]
LGTM! Huge thanks for your efforts :)

I'm not going to wait for another review, I'm pretty confident it's all good.

**Comment by monsieurtanuki** [comment]
Thank you @JaffaKetchup for your reviews and comments!
Just an additional comment, as there are breaking changes anyway: perhaps the "spacing" parameter of "dotted" could be a size in pixels rather than a factor to apply on top of paint size.

**Pull request merged**
